src/IRVisualization.cpp

```diff
@@ -0,0 +1,1585 @@
+}
+
+string IRVisualization::read_write_table(StmtSize &size) const {
+    stringstream read_write_table_ss;
+
+    // open table
+    read_write_table_ss << "<table class='costTable'>";
+
+    // Prod | Cons
+    read_write_table_ss << "<tr>";
+
+    read_write_table_ss << "<th colspan='2' class='costTableHeader middleCol'>";
+    read_write_table_ss << "Written";
+    read_write_table_ss << "</th>";
+
+    read_write_table_ss << "<th colspan='2' class='costTableHeader'>";
+    read_write_table_ss << "Read";
+    read_write_table_ss << "</th>";
+
+    read_write_table_ss << "</tr>";
+
+    // produces and consumes are empty
+    if (size.empty()) {
+        internal_error << "\n\n"
+                       << "IRVisualization::read_write_table - size is empty"
+                       << "\n";
+    }
+
+    // produces and consumes aren't empty
+    else {
+        vector<string> rows;
+
+        // fill in producer variables
+        for (const auto &produce_var : size.writes) {
+            string ss;
+            ss += "<td class='costTableData'>";
+            ss += produce_var.first + ": ";
+            ss += "</td>";
+
+            ss += "<td class='costTableData middleCol'>";
+            ss += produce_var.second;
+            ss += "</td>";
+
+            rows.push_back(ss);
+        }
+
+        // fill in consumer variables
+        unsigned long row_num = 0;
+        for (const auto &consume_var : size.reads) {
+            string ss;
+            ss += "<td class='costTableData'>";
+            ss += consume_var.first + ": ";
+            ss += "</td>";
+
+            ss += "<td class='costTableData'>";
+            ss += consume_var.second;
+            ss += "</td>";
+
+            if (row_num < rows.size()) {
+                rows[row_num] += ss;
+            } else {
+                // pad row with empty cells for produce
+                string s_empty;
+                s_empty += "<td colspan='2' class='costTableData middleCol'>";
+                s_empty += "</td>";
+
+                rows.push_back(s_empty + ss);
+            }
+            row_num++;
+        }
+
+        // pad row with empty calls for consume
+        row_num = size.reads.size();
+        while (row_num < size.writes.size()) {
+            string s_empty;
+            s_empty += "<td class='costTableData'>";
+            s_empty += "</td>";
+            s_empty += "<td class='costTableData'>";
+            s_empty += "</td>";
+
+            rows[row_num] += s_empty;
+            row_num++;
+        }
+
+        // add rows to read_write_table_ss
+        for (const auto &row : rows) {
+            read_write_table_ss << "<tr>";
+            read_write_table_ss << row;
+            read_write_table_ss << "</tr>";
+        }
+    }
+
+    // close table
+    read_write_table_ss << "</table>";
+
+    return read_write_table_ss.str();
+}
+string IRVisualization::allocate_table(vector<string> &allocation_sizes) const {
+    stringstream allocate_table_ss;
```

Comment:
nit, here and elsewhere: probably prefer `ostringstream` for a stream that is only written to